mesonbuild/interpreter.py

```diff
@@ -3409,12 +3409,16 @@ def func_configure_file(self, node, args, kwargs):
             raise InterpreterException('@INPUT@ used as command argument, but no input file specified.')
         # Validate output
         output = kwargs['output']
+        ofile_rpath = os.path.join(self.subdir, output)
```

Comment:
This will give the wrong answer if --layout=flat is being used (or if the backend is one which only supports that)

I think there is some interface for asking the backend the path for this output, but it might be a bit under-developed at the moment.

Reply:
Well, I guess this is were some documentation for developers comes in handy :) I was unaware of the flat layout.

On the other hand, as far as I can see this is the way the path to the output file is build. See a few lines below:
```python
    if ifile_abs:
        values = mesonlib.get_filenames_templates_dict([ifile_abs], None)
        outputs = mesonlib.substitute_values([output], values)
        output = outputs[0]
    if os.path.dirname(output) != '':
        raise InterpreterException('Output file name must not contain a subdirectory.')
    (ofile_path, ofile_fname) = os.path.split(os.path.join(self.subdir, output))
```
As far as I can tell none of the calls to the functions in `mesonlib` touches the path.

If someone with more knowledge could chime in and suggest what to do, I will happily implement it. 

What would happen in the flat case, would the files be overwritten, even in different subdirs?Should I only check for the filename and warn that they could be overwritten?

Reply:
> What would happen in the flat case, would the files be overwritten, even in different subdirs? Should I only check for the filename and warn that they could be overwritten?

There should be a helper function somewhere for getting the path to a file when using any layout, which will reduce to the basename when using flat layout. You shouldn't always warn based on the basename of the file, we generally don't force the lowest-common-denominator on all configurations/platforms.

Also note that the flat layout is not very well-used or maintained (it's not tested at all on the CI), and as soon as we have a way to run applications without extra configuration on Windows (`meson run`, or similar) we will start to seriously consider removing it.

Reply:
After reading the code again and running a test ^^ I can say that configuration files ignore the flat layout. Even with a flat layout the configuration files are written into subdirs.

Reply:
> @nirbheek There are two tests with layout=flat

I meant that we don't run the full test suite with `--layout=flat` (lots of tests would fail), in the same way that we run with `--unity=on`.